python/ray/data/read_api.py

```diff
@@ -338,11 +338,15 @@ def read_datasource(
     block_list.compute_first_block()
     block_list.ensure_metadata_for_first_block()
```

Comment:
IMO we should get rid of this default behavior of always computing the first block and ensuring the metadata for the first block, and instead make the dataset fully lazy by default. We can still progressively launch read tasks to e.g. fetch the schema, see the first few rows, streaming iteration directly on the read, etc. We'd just move that logic into `ExecutionPlan.schema()`.

Reply:
Hmm, that seems fine as long as we still resolve the metadata for parquet and so on. This would only apply to JSON/CSV presumably. Maybe we can also add an optimized schema resolver for these file types that peeks at the header of the file only.

Reply:
Yes agreed, we shall improve the schema resolver in the long term. I am 100% agree to get rid of always computing the first block. Do we want to do it in a separate PR? https://github.com/ray-project/ray/pull/31286#issuecomment-1372976970